thcrap_tsa/src/anm.cpp

```diff
@@ -1054,18 +1295,106 @@ int patch_anm(void *file_inout, size_t size_out, size_t size_in, const char *fn,
 			for(const auto &sprite : entry.sprites) {
 				bounds_draw_rect(bounds, entry.x, entry.y, sprite);
 			}
-			// Do the patching
-			stack_game_png_apply(entry);
+
+			if (entry.thtxoffset) {
+				// Do the patching
+				thtx_header_t* thtx_in = (thtx_header_t*)(anm_entry_in + entry.thtxoffset);
+
+				img_patch_t img_patch = img_get_patch(entry, thtx_in);
+				stack_game_png_apply(img_patch, entry.sprites);
+				img_encode(img_patch);
+
+				thtx_header_t* thtx_out = (thtx_header_t*)(anm_entry_out + entry.thtxoffset);
+				thtx_out->size = img_patch.img_size;
+
+				memcpy(thtx_out->data, img_patch.img_data, img_patch.img_size);
+
+				uint32_t* nextoffset_ptr = anm_get_nextoffset_ptr(anm_entry_out);
+				size_t out_nextoffset = entry.thtxoffset + sizeof(thtx_header_t) + img_patch.img_size;
+				if (*nextoffset_ptr)
+					*nextoffset_ptr = out_nextoffset;
+				anm_entry_out += out_nextoffset;
+
+				free(img_patch.img_data);
+			}
+			else {
+				goto next_no_thtx;
+			}
+		}
+		else if (entry.next) {
+next_no_thtx:
+			anm_entry_out += entry.next - anm_entry_in;
 		}
 		if(!entry.next) {
 			bounds_store(name_prev, bounds);
 		}
-		anm_entry_out = entry.next;
+		anm_entry_in = entry.next;
 		anm_entry_clear(entry);
 	}
 	png_image_clear(bounds);
 	png_image_clear(png);
+	free(anm_entry_in);
 
 	log_debugf("-------------\n");
 	return 1;
 }
+
+template<typename T>
+inline size_t anm_get_imgsize_diff(void* anm_file, size_t anm_size) {
+	auto* entry = (T*)anm_file;
+	size_t ret = 0;
+	while ((size_t)entry - (size_t)anm_file < anm_size) {
+		if (entry->thtxoffset) {
+			thtx_header_t* thtx = (thtx_header_t*)((uintptr_t)entry + entry->thtxoffset);
+			ret += (thtx->w * thtx->h * 4) - thtx->size;
+		}
+
+		if (entry->nextoffset) {
+			entry = (T*)((uintptr_t)entry + entry->nextoffset);
+		}
+		else {
+			break;
+		}
+	}
+	return ret;
+}
+
+size_t anm_get_size(const char* fn, json_t* patch, size_t patch_size) {
+	tlnote_remove();
```

Comment:
Is it really needed?

Reply:
Before, the `patch_size_func` parameter passed into `patchhook_register` for `*.anm` was a function was `tlnote_remove_size_hook`, which is just
```cpp
size_t tlnote_remove_size_hook(const char *fn, json_t *patch, size_t patch_size)
{
	// We want these gone as early as possible
	tlnote_remove();
	return 0;
}
```

Reply:
Okay, makes sense. I missed that part in the diff.